x/wasm/keeper/snapshotter.go

```diff
@@ -0,0 +1,172 @@
+	return &WasmSnapshotter{
+		wasm: wasm,
+		cms:  cms,
+	}
+}
+
+func (ws *WasmSnapshotter) SnapshotName() string {
+	return "WASM Code Snapshot"
+}
+
+func (ws *WasmSnapshotter) SnapshotFormat() uint32 {
+	return 1
+}
+
+func (ws *WasmSnapshotter) SupportedFormats() []uint32 {
+	return []uint32{1}
+}
+
+func (ws *WasmSnapshotter) Snapshot(height uint64, protoWriter protoio.Writer) error {
+	var rerr error
+	ctx := sdk.NewContext(ws.cms, tmproto.Header{}, false, log.NewNopLogger())
+
+	seen := make(map[string]bool)
+
+	ws.wasm.IterateCodeInfos(ctx, func(id uint64, info types.CodeInfo) bool {
+		// Many code ids may point to the same code hash... only sync it once
+		hexHash := hex.EncodeToString(info.CodeHash)
+		// if seen, just skip this one and move to the next
+		if seen[hexHash] {
+			return false
+		}
+		seen[hexHash] = true
+
+		// load code and abort on error
+		wasmBytes, err := ws.wasm.GetByteCode(ctx, id)
+		if err != nil {
+			rerr = err
+			return true
+		}
+
+		compressedWasm, err := ioutils.GzipIt(wasmBytes)
+		if err != nil {
+			rerr = err
+			return true
+		}
+
+		err = snapshot.WriteExtensionItem(protoWriter, compressedWasm)
+		if err != nil {
+			rerr = err
+			return true
+		}
+
+		return false
+	})
+
+	return rerr
+}
+
+func (ws *WasmSnapshotter) Restore(
+	height uint64, format uint32, protoReader protoio.Reader,
+) (snapshot.SnapshotItem, error) {
+	if format == 1 {
+		return ws.processAllItems(height, protoReader, restoreV1, finalizeV1)
+	}
+	return snapshot.SnapshotItem{}, snapshot.ErrUnknownFormat
+}
+
+func restoreV1(ctx sdk.Context, k Keeper, compressedCode []byte) error {
+	wasmCode, err := ioutils.Uncompress(compressedCode, k.GetMaxWasmCodeSize(ctx))
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrCreateFailed, err.Error())
+	}
+
+	// FIXME: check which codeIDs the checksum matches??
+	_, err = k.wasmVM.Create(wasmCode)
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrCreateFailed, err.Error())
+	}
+	return nil
+}
+
+func finalizeV1(ctx sdk.Context, k Keeper) error {
+	// FIXME: ensure all codes have been uploaded?
+	return k.InitializePinnedCodes(ctx)
+}
+
+func (ws *WasmSnapshotter) processAllItems(
+	height uint64,
+	protoReader protoio.Reader,
+	cb func(sdk.Context, Keeper, []byte) error,
+	finalize func(sdk.Context, Keeper) error,
+) (snapshot.SnapshotItem, error) {
+	ctx := sdk.NewContext(ws.cms, tmproto.Header{}, false, log.NewNopLogger())
+
+	for {
+		item := snapshot.SnapshotItem{}
+		err := protoReader.ReadMsg(&item)
+		if err == io.EOF {
+			return snapshot.SnapshotItem{}, finalize(ctx, ws.wasm)
+		} else if err != nil {
+			return snapshot.SnapshotItem{}, sdkerrors.Wrap(err, "invalid protobuf message")
+		}
+
+		// if it is not another ExtensionPayload message, then it is not for us.
+		// we should return it an let the manager handle this one
+		payload := item.GetExtensionPayload()
+		if payload == nil {
+			return item, nil
```

Comment:
@yihuang If I understand https://github.com/CosmWasm/wasmd/pull/816#discussion_r859703173 correctly, then this is the correct behavior. Can you confirm or correct me?

Reply:
Got you, I need to run finalise here as well.